### DISCUSSION TITLE
Can't get cross platform to work except for Windows

### DISCUSSION THREAD

**Discussion by Yveske1966** [AvaloniaUI/Avalonia#13726] [Q&A] [answered]
I'm coming over from CBuilder, having a showstopper with Maui and came across Avalon which is supposed to be production ready...
I'm running on Windows 11 Pro 22H2, AMD Ryzen 9 3950X with 64GB using Visual Studio Community 2022 (64-bit) - Preview 17.9.0

Is there a guide with the correct setups of the different platforms, because I can't get the default program to run.
What are the correct setups for Android, IOS, MacOS and Linux projects ?
- Compiling for Android complains the program doesn't contain a static 'Main' method. Also how do I specify emulator or device ?
- IOS compiles as console app (?) but remote device asks me to select a valid device first -> Where ? Using Ipad mini3 with 12.5.7
  Using the emulator I get an error "GetFileSystemEntries" from the MacOs project tree and a "Clang++ exited with code1:"
- The MacOS project is a copy of the Desktop project with target OS set to macOS (as specified in the default desktop project).
  It also compiles as a console app (?), but trying to run results in "Unable to start debugging. The startup project cannot be launched..."
  The startup object is set to AvaloniaApplication1.Desktop.MacOS.program and I'm connected to an M1 Mac Mini with Sonoma 14.1.1
  Trying to make a release version, Pack complains runtimeconfig.json is not found
- The Linux project is also a copy of the Desktop project, but here I can't specify Linux as target OS. How to proceed ?
  I want to use Raspberry Pi.

Hoping someone can point me in the right direction, TIA
Yveske

**Comment by thevortexcloud** [comment]
>Compiling for Android complains the program doesn't contain a static 'Main' method. 

Have you installed the android workload? Try `dotnet workload install android` in a terminal. 

> Also how do I specify emulator or device ?

This really depends on your IDE/setup. In Rider a device option should appear in the run options when you plug one in or connect via ADB.

> The Linux project is also a copy of the Desktop project, but here I can't specify Linux as target OS. How to proceed ?
I want to use Raspberry Pi.

There is no Linux target. Microsoft has not made one. All code for Linux in .NET is "crossplatform" as far as the runtime is concerned. You need to do runtime checks (EG using [`OperatingSystem.IsLinux()`](https://learn.microsoft.com/en-us/dotnet/api/system.operatingsystem.islinux?view=net-7.0)) or write your own compile time checks if you want to use Linux only code. In some ways this is better since it means you can easily share code with other targets since as far as the runtime is concerned it's all compatible. But it also means Microsoft has not exposed any Linux specific APIs like you would get on Windows/Mac with each of their platform TFMs. Which means you need to write all the native interop yourself if you need access to things like X11 APIs.

For Avalonia specifically though there are multiple backends that can be used on Linux. The default is X11 if you want to run it in a full GUI environment. But otherwise you can use FBDev or DRM. There is a short guide [here](https://docs.avaloniaui.net/docs/next/guides/platforms/rpi/running-on-raspbian-lite-via-drm) showing using the DRM backend on a Raspberry Pi. But that may not be desirable depending on what you are making as it will make your app have exclusive control over the display.

I can't really help with any of the Mac/iOS stuff as I don't use or have either of them.

In any case, have you tried the default cross platform template? The Desktop project will run on Mac, Linux and Windows out of the box. If you don't need access to any of the platform specific APIs it works fine.

EDIT: I also suggest reading the .NET TFM page to get a better understanding of how all the platform targets work,

https://learn.microsoft.com/en-us/dotnet/standard/frameworks

**Reply by Yveske1966** [reply]
Thanks for your answer !
Yes, I do have all necessary workloads installed.
Seems I have some more reading to do first thing next week !
Will come back and list the correct steps for future reference and mark your comment as answer if it gets me there.
Thanks again !

**Reply by thevortexcloud** [reply]
> Yes, I do have all necessary workloads installed.

Hmm strange. I have seen things like this happen before when using third party builds of .NET (I am still not clear why they don't work with some workloads) such as those provided in a lot of Linux distros. But I doubt that is what is happening if you are using Visual Studio. I however have never tried to build an android app on Windows with .NET.

>Seems I have some more reading to do first thing next week !

If you want even more reading to do about how to handle cross platform Avalonia code in a less awful way than what Microsoft wants people to do (which is using preprocessor directives) while using OS specific TFMs. Look into how to use Dependency injection, or [read my comment](https://github.com/AvaloniaUI/Avalonia/discussions/13455#discussioncomment-7464329) I wrote showing a very minimal example about how to use it with Avalonia and multiple platforms.

Microsoft also has [this](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-8.0)  somewhat overdetailed article explaining the concepts of Dependency injection. However it's mainly targeted at ASP Core developers. But the concepts and most of the suggestions about how to use it are the same.

**Comment by Yveske1966** [comment, marked as answer]
I'm sorry to have bothered all of you, the simple answer is **Avalonia doesn't like spaces in the name of the solution**.

**Reply by thevortexcloud** [reply]
No problem. .NET in general does not like spaces in most places. Visual Studio usually puts underscores in if you try. But the general convention is to use `.` so it forms part of the namespace properly. If Avalonia specifically can't handle it and a console app can, that actually sounds like a possible bug in Avalonia. But it's not something I have tried as it's a fairly weird thing to do in the .NET world.

**Reply by Yveske1966** [reply]
I'm new to dotnet and had no trouble with it in Maui, so I hadn't given it any thought in Avalonia.
Hoping I now can make short work of porting my CBuilder app to dotnet as CBuilder is a dead end for multi-platform.
Thanks again for your support !

**Reply by thevortexcloud** [reply]
> so I hadn't given it any thought in Avalonia.

Might be worth opening an issue then if you feel like it.

> Thanks again for your support !

You are welcome! Feel free to open a new discussion if you need further assistance. One small tip though: Avalonia is _very_ similar to other XAML frameworks such as WPF and UWP. If you are struggling to find an Avalonia specific piece of documentation or examples, try searching for one for one of those. ~60% of the time you will find something that works with Avalonia and some minimal tweaking. 

I also suggest using MVVM if you can. It makes things much easier when using XAML.

**Reply by Yveske1966** [reply]
Never used WPF or UWP, but I was "nearing completion" in Maui, where I was already using MVVM with dependency injection.
I guess that will make most of my code more or less transferable and I only need to make myself acquainted with the UI stuff.

It's definitely a bug, because while I don't mind Avalonia (like Maui) replaces it with underscores in some places, in Avalonia it doesn't change the actual path as well so it can't find its own files no longer. Will make a bug report about it.

**Reply by timunie** [reply]
@thevortexcloud in a console App you have to wrap the path in `"My Path With Spaces\..."`. Also how paths are handled heavily depends on the OS. So while Windows may work, Linux may fail. Moreover I believe one cannot compile the Mac or iOS App without an actual Mac. Not our fault, but a restriction that Apple puts on it.

**Reply by Yveske1966** [reply]
![My program with spaces](https://github.com/AvaloniaUI/Avalonia/assets/68047793/c7614043-01b7-407e-9ce4-6bedabc859c1)

**Reply by thevortexcloud** [reply]
>Also how paths are handled heavily depends on the OS. 

Sort of. Linux certainly has far more legal characters and file names than Windows does for paths. EG it's possible to create files on Linux Windows will refuse to delete when viewed on Windows. Most Linux distros are also configured with case sensitive paths. But `"My file.csproj"` will be just as valid on Window as it is on Linux (although a lot of programs on Linux will die very quickly if you try to use a space, which makes it not worth it even if it is valid). In any case, this is why .NET has `Path.Combine` and friends to handle all of this for you. 

Anyway, I only mentioned trying a console app to see if it would build as I have never tried doing it with pure spaces before. I just tried it myself on Linux doing a console app using the CLI and it works fine. I have not tried Avalonia though.
